skills/contextual-embeddings/guide.ipynb

```diff
@@ -0,0 +1,1372 @@
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@5: 86.37%\n",
+      "Total Score: 0.8637192780337941\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Evaluating retrieval: 100%|██████████| 248/248 [00:06<00:00, 40.05it/s]\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@10: 92.81%\n",
+      "Total Score: 0.9280913978494625\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Evaluating retrieval: 100%|██████████| 248/248 [00:06<00:00, 39.64it/s]"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@20: 93.78%\n",
+      "Total Score: 0.9378360215053763\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "r5 = evaluate_db(contextual_db, 'data/evaluation_set.jsonl', 5)\n",
+    "r10 = evaluate_db(contextual_db, 'data/evaluation_set.jsonl', 10)\n",
+    "r20 = evaluate_db(contextual_db, 'data/evaluation_set.jsonl', 20)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Contextual BM25\n",
+    "\n",
+    "Contextual embeddings is an improvement on traditional semantic search RAG, but we can improve performance further. In this section we'll show you how you can use contextual embeddings and *contextual* BM25 together. While you can see performance gains by pairing these techniques together without the context, adding context to these methods reduces the top-20-chunk retrieval failure rate by 42%.\n",
+    "\n",
+    "We'll use elastic search for the BM25 portion of this section, which will require you to have the elasticsearch library installed and it will also require you to spin up an Elasticsearch server in the background. The easiest way to do this is to install [docker](https://docs.docker.com/engine/install/) and run the following docker command:\n",
+    "\n",
+    "`docker run -d --name elasticsearch -p 9200:9200 -p 9300:9300 -e \"discovery.type=single-node\" -e \"xpack.security.enabled=false\" elasticsearch:8.8.0`\n",
+    "\n",
+    "One difference between a typical BM25 search and what we'll do in this section is that, for each chunk, we'll run each BM25 search on both the chunk content and the additional context that we generated in the previous section. From there, we'll use a technique called reciprocal rank fusion to merge the results from our BM25 search with our semantic search results. This allows us to perform a hybrid search across both our BM25 corpus and vector DB to return the most optimal documents for a given query.\n",
+    "\n",
+    "### Note:\n",
```

Comment:
nit: Maybe for notes (here and above), just bold the **Note:** and then have the text be after. Seems strange to have a new header for such a small note. lol

Reply:
also, might be worth elaborating on what the weightings do.